rust-src/concordium_base/src/web3id/v1/proofs.rs

```diff
@@ -0,0 +1,1132 @@
+            statement.verify(
+                ProofVersion::Version2,
+                global_context,
+                transcript,
+                cmm_attributes,
+                proof,
+            )
+        })
+    })
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> AccountBasedSubjectClaims<C, AttributeType> {
+    /// Proof described in "15.4.1 Account Based Presentations" (blue paper v2.2.0)
+    fn prove<P: Pairing<ScalarField = C::Scalar>>(
+        self,
+        global_context: &GlobalContext<C>,
+        transcript: &mut RandomOracle,
+        csprng: &mut (impl Rng + CryptoRng),
+        now: chrono::DateTime<chrono::Utc>,
+        input: CredentialProofPrivateInputs<P, C, AttributeType>,
+    ) -> Result<AccountBasedCredentialV1<C, AttributeType>, ProofError> {
+        let CredentialProofPrivateInputs::Account(AccountCredentialProofPrivateInputs {
+            attribute_values: values,
+            attribute_randomness: randomness,
+            issuer,
+        }) = input
+        else {
+            return Err(ProofError::CommitmentsStatementsMismatch);
+        };
+
+        let statement_proofs = prove_statements(
+            &self.statements,
+            values,
+            randomness,
+            global_context,
+            transcript,
+            csprng,
+        )?;
+
+        Ok(AccountBasedCredentialV1 {
+            proof: ConcordiumZKProof {
+                created_at: now,
+                proof: AccountCredentialProofs { statement_proofs },
+                proof_type: ConcordiumProofType::ConcordiumZKProofV4,
+            },
+            subject: AccountCredentialSubject {
+                cred_id: self.cred_id,
+                statements: self.statements,
+                network: self.network,
+            },
+            issuer,
+        })
+    }
+}
+
+impl<C: Curve, AttributeType: Attribute<C::Scalar>> IdentityBasedSubjectClaims<C, AttributeType> {
+    /// Proof described in "15.4.2 Identity Based Credential" (blue paper v2.2.0)
+    fn prove<P: Pairing<ScalarField = C::Scalar>>(
+        self,
+        global_context: &GlobalContext<C>,
+        transcript: &mut RandomOracle,
+        csprng: &mut (impl Rng + CryptoRng),
+        now: chrono::DateTime<chrono::Utc>,
+        input: CredentialProofPrivateInputs<P, C, AttributeType>,
+    ) -> Result<IdentityBasedCredentialV1<P, C, AttributeType>, ProofError> {
+        let CredentialProofPrivateInputs::Identity(IdentityCredentialProofPrivateInputs {
+            ip_context,
+            id_object,
+            id_object_use_data,
+        }) = input
+        else {
+            return Err(ProofError::CommitmentsStatementsMismatch);
+        };
+
+        let attributes_handling: BTreeMap<_, _> = self
+            .statements
+            .iter()
+            .map(|stmt| match stmt {
+                AtomicStatement::RevealAttribute { statement } => {
+                    (statement.attribute_tag, IdentityAttributeHandling::Commit)
```

Comment:
Maybe worth a comment that `IdentityAttributeHandling::Reveal` is relevant only when the `attribute` is revealed on-chain.